### PR TITLE
fix: change the default hierarchical facet behavior

### PR DESCRIPTION
**Pull request by vvo**
To ease navigation, we always:
  - provide the root level categories
  - provide the current refinement sub categories (`fruits > citrus > *`:
  n + 1)
  - provide the parent refinement (`fruits > citrus` => `fruits`: n -1)
  categories
  - refine the search using the current hierarchical refinement

### CODE REVIEW

**Comment by vvo** [comment]
I do think that we are good for helper release after this one :)

**Comment by vvo** [comment]
So tomorrow morning

**Comment by bobylito** [comment]
Nice looking at that right away :)

**Comment by vvo** [comment]
Not fan of a lot of code in it, needs refactor at some point

**Comment by vvo** [comment]
Documentation for hierarchical facet:

https://github.com/algolia/algoliasearch-helper-js/tree/f7cea4938e70b7d509474ed851728f1530f2d08d#hierarchical-facets

File: src/SearchResults/generate-hierarchical-tree.js [hunk 1]
```diff
@@ -17,9 +17,8 @@ function generateTrees(state) {
       state.hierarchicalFacetsRefinements[hierarchicalFacet.name][0] || '';
     var hierarchicalSeparator = state._getHierarchicalFacetSeparator(hierarchicalFacet);
     var sortBy = prepareHierarchicalFacetSortBy(state._getHierarchicalFacetSortBy(hierarchicalFacet));
-    var alwaysGetRootLevel = hierarchicalFacet.alwaysGetRootLevel;
 
-    return reduce(hierarchicalFacetResult, generateHierarchicalTree(sortBy, hierarchicalSeparator, hierarchicalFacetRefinement, alwaysGetRootLevel), {
+    return reduce(hierarchicalFacetResult, generateHierarchicalTree(sortBy, hierarchicalSeparator, hierarchicalFacetRefinement), {
```
Comment (bobylito) [inline src/SearchResults/generate-hierarchical-tree.js]: `generateHierarchicalTree(sortBy, hierarchicalSeparator, hierarchicalFacetRefinement)` could be put in a variable to ease readability and shorten line length?

File: src/SearchResults/generate-hierarchical-tree.js [hunk 2]
```diff
@@ -58,7 +57,7 @@ function generateHierarchicalTree(sortBy, hierarchicalSeparator, currentRefineme
           //
           // If parent refinement is `beers`, then we do not want to have `bières > Belges`
           // showing up
-          pick(hierarchicalFacetResult.data, filterFacetValues(parent.path, currentRefinement, hierarchicalSeparator, alwaysGetRootLevel)),
+          pick(hierarchicalFacetResult.data, filterFacetValues(parent.path, currentRefinement, hierarchicalSeparator)),
```
Comment (bobylito) [inline src/SearchResults/generate-hierarchical-tree.js]: Variable here could be nice too.

File: src/SearchResults/index.js [hunk 3]
```diff
@@ -239,6 +239,7 @@ function SearchResults(state, algoliaResponse) {
 
   var facetsIndices = getIndices(state.facets);
   var disjunctiveFacetsIndices = getIndices(state.disjunctiveFacets);
+  var disjunctiveIndexResult;
```
Comment (bobylito) [inline src/SearchResults/index.js]: Set an initial value here could help read the code (we know the type of the variable).

**Comment by bobylito** [comment]
The code is complex but powerful is the feature! I'll let you make the changes (if you accept them) but otherwise seems good.

**Comment by vvo** [comment]
@bobylito Made all the changes, thanks for review!

**Pull request merged**